File: net.akehurst.language/processor/src/test/java/net/akehurst/language/processor/vistraq/test_QueryParserValid.java
package net.akehurst.language.processor.vistraq;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameter;
import org.junit.runners.Parameterized.Parameters;

import net.akehurst.language.core.analyser.UnableToAnalyseExeception;
import net.akehurst.language.core.grammar.GrammarRuleNotFoundException;
import net.akehurst.language.core.grammar.Grammar;
import net.akehurst.language.core.parser.ParseFailedException;
import net.akehurst.language.core.parser.ParseTreeException;
import net.akehurst.language.core.sppt.SharedPackedParseTree;
import net.akehurst.language.processor.LanguageProcessor;
import net.akehurst.language.processor.OGLanguageProcessor;

@RunWith(Parameterized.class)
public class test_QueryParserValid {

    static class Data {
        public Data(final String queryStr) {
            this.queryStr = queryStr;
        }

        public final String queryStr;

        // --- Object ---
        @Override
        public String toString() {
            return this.queryStr;
        }
    }

    @Parameters(name = "{0}")
    public static Collection<Object[]> data() throws IOException {
        final List<Object[]> col = new ArrayList<>();
        final InputStream is = test_QueryParserInValid.class.getResourceAsStream("/vistraq/sampleValidQueries.txt");
        final BufferedReader br = new BufferedReader(new InputStreamReader(is));
        String line = br.readLine();
        while (null != line) {
            col.add(new Object[] { new Data(line) });
            line = br.readLine();
        }
        return col;
    }

    @Parameter
    public Data data;

    private LanguageProcessor processor;

    private String toString(final SharedPackedParseTree tree) {
        String res = tree.asString().replaceAll(System.lineSeparator(), " ");
        res = res.trim();
        return res;
    }

    @Before
    public void setup() {
        try {
            final OGLanguageProcessor oglProc = new OGLanguageProcessor();
            final InputStreamReader reader = new InputStreamReader(test_QueryParserInValid.class.getResourceAsStream("/vistraq/Query.ogl"));
            final Grammar grammar = oglProc.process(reader, "grammarDefinition", Grammar.class);
            this.processor = new LanguageProcessor(grammar, null);
        } catch (final ParseFailedException e) {
            Assert.fail("Parsing query grammar failed " + e.getLongestMatch());
        } catch (final UnableToAnalyseExeception e) {
            Assert.fail("Analysing query grammar failed");
        }
    }

    @Test
    public void test() throws ParseFailedException, ParseTreeException, GrammarRuleNotFoundException {

        final String queryStr = this.data.queryStr;
        final SharedPackedParseTree result = this.processor.getParser().parse("query", queryStr);
        Assert.assertNotNull(result);
        final String resultStr = this.toString(result);
        Assert.assertEquals(queryStr, resultStr);

    }

}

File: net.akehurst.language/agl-processor/src/jvm8Test/kotlin/processor/vistraq/test_QueryParserValid.kt
/**
 * Copyright (C) 2020 Dr. David H. Akehurst (http://dr.david.h.akehurst.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.akehurst.language.agl.processor.vistraq

import net.akehurst.language.agl.Agl
import net.akehurst.language.agl.GrammarString
import net.akehurst.language.parser.leftcorner.ParseOptionsDefault
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.test.assertNotNull
import kotlin.test.assertTrue


@RunWith(Parameterized::class)
class test_QueryParserValid(val data: Data) {

    private companion object {

        val grammarStr = test_QueryParserValid::class.java.getResource("/vistraq/version_/grammar.agl")?.readText() ?: error("File not found")

        //val grammarStr = test_QueryParserValid::class.java.getResource("/vistraq/Query.agl")?.readText() ?: error("File not found")
        var processor = tgqlprocessor()

        var sourceFiles = arrayOf("/vistraq/version_/valid/examples.txt")

        fun tgqlprocessor() = Agl.processorFromStringSimple(GrammarString(grammarStr)).processor!!

        @JvmStatic
        @Parameters(name = "{0}")
        fun data(): Collection<Array<Any>> {
            val col = ArrayList<Array<Any>>()
            for (sourceFile in sourceFiles) {
                // val inps = ClassLoader.getSystemClassLoader().getResourceAsStream(sourceFile)
                val inps = test_QueryParserValid::class.java.getResourceAsStream(sourceFile) ?: error("File not found")

                val br = BufferedReader(InputStreamReader(inps))
                var line: String? = br.readLine()
                while (null != line) {
                    line = line.trim { it <= ' ' }
                    if (line.isEmpty()) {
                        // blank line
                        line = br.readLine()
                    } else if (line.startsWith("//")) {
                        // comment
                        line = br.readLine()
                    } else {
                        col.add(arrayOf(Data(sourceFile, line)))
                        line = br.readLine()
                    }
                }
            }
            return col
        }
    }

    class Data(val sourceFile: String, val queryStr: String) {

        // --- Object ---
        override fun toString(): String {
            return this.sourceFile + ": " + this.queryStr
        }
    }

    @Test(timeout = 2000)
    fun parse() {
        val queryStr = this.data.queryStr
        val goal = "query"
        val result = processor.parse(queryStr, ParseOptionsDefault(goal))
        assertTrue(result.issues.errors.isEmpty(), result.issues.toString())
        assertNotNull(result.sppt)
        val resultStr = result.sppt!!.asSentence
        Assert.assertEquals(queryStr, resultStr)
    }

    @Test(timeout = 2000)
    fun process() {
        val queryStr = this.data.queryStr
        val goal = "query"
        val result = processor.process(queryStr, Agl.options { parse { goalRuleName(goal) } })
        assertTrue(result.issues.errors.isEmpty(), result.issues.toString())
        assertNotNull(result.asm)
    }
}
